crates/milli/src/vector/json_template/mod.rs

```diff
@@ -0,0 +1,283 @@
+    fn clone(&self) -> Self {
+        Self::new(self.value.clone()).unwrap()
+    }
+}
+
+struct TemplateAtPath {
+    template: Template,
+    path: ValuePath,
+}
+
+impl std::fmt::Debug for TemplateAtPath {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TemplateAtPath")
+            .field("template", &&"template")
+            .field("path", &self.path)
+            .finish()
+    }
+}
+
+/// Error that can occur either when parsing the templates in the value, or when trying to render them.
+#[derive(Debug)]
+pub struct Error {
+    template_error: liquid::Error,
+    path: ValuePath,
+}
+
+impl Error {
+    /// Produces an error message when the error happened at rendering time.
+    pub fn rendering_error(&self, root: &str) -> String {
+        format!(
+            "in `{}`, error while rendering template: {}",
+            path_with_root(root, self.path.iter()),
+            &self.template_error
+        )
+    }
+
+    /// Produces an error message when the error happened at parsing time.
+    pub fn parsing(&self, root: &str) -> String {
+        format!(
+            "in `{}`, error while parsing template: {}",
+            path_with_root(root, self.path.iter()),
+            &self.template_error
+        )
+    }
+}
+
+impl JsonTemplate {
+    /// Creates a new `JsonTemplate` by parsing all strings inside the value as templates.
+    ///
+    /// # Error
+    ///
+    ///  - If any of the strings contains a template that cannot be parsed.
+    pub fn new(value: Value) -> Result<Self, Error> {
+        let templates = build_templates(&value)?;
+        Ok(Self { value, templates })
+    }
+
+    /// Renders this value by replacing all its strings with the rendered version of the template they represent from the given context.
+    ///
+    /// # Error
+    ///
+    /// - If any of the strings contains a template that cannot be rendered with the given context.
+    pub fn render(&self, context: &dyn liquid::ObjectView) -> Result<Value, Error> {
+        let mut rendered = self.value.clone();
+        for TemplateAtPath { template, path } in &self.templates {
+            let injected_value =
+                template.render(context).map_err(|err| error_with_path(err, path.clone()))?;
+            inject_value(&mut rendered, path, Value::String(injected_value));
+        }
+        Ok(rendered)
+    }
+
+    /// Renders this value by replacing all its strings with the rendered version of the template they represent from the contents of the given document.
+    ///
+    /// # Error
+    ///
+    /// - If any of the strings contains a template that cannot be rendered with the given document.
+    pub fn render_document<'a, 'doc: 'a, D: Document<'a> + std::fmt::Debug>(
+        &'a self,
+        document: D,
+        doc_alloc: &'doc Bump,
+    ) -> Result<Value, Error> {
+        let document = ParseableDocument::new(document, doc_alloc);
+        let v: Vec<u32> = vec![];
```

Comment:
will create a `fields` field, but it will be empty!

- consider removing the `fields` field or to populate it